target/linux/ramips/image/mt7620.mk

```diff
@@ -488,6 +488,25 @@ define Device/hnet_c108
 endef
 TARGET_DEVICES += hnet_c108
 
+define Device/hootoo_ht-tm05
+  SOC := mt7620n
+  IMAGE_SIZE := 6144k
+  DEVICE_VENDOR := HooToo
+  DEVICE_MODEL := HT-TM05
+  DEVICE_PACKAGES := kmod-usb2 kmod-usb-ohci kmod-i2c-ralink
+  LOADER_TYPE := bin
+  LOADER_FLASH_OFFS := 0xFD200000
```

Comment:
This results in an overflow in loader:
Our current loader has an incorrect AR71XX_FLASH_START at 0x1f000000. After it's converted to kseg1 address for MIPS, it becomes 0xbf000000.
This code told it to search from AR71XX_FLASH_START + FLASH_OFFS which is 0x1BC200000. It happened to be the needed 0xbc200000 offset after dropping the overflowed 32rd bit.

For the sake of correctness, you should add a separated commit fixing FLASH_START and FLASH_END to be 0x1c000000 and 0x1d000000 respectively, and define FLASH_OFFS as 0x200000 here.

Reply:
Hi @981213 , thank you for joining the review!

`FLASH_END` is unused in `loader.c`, but `FLASH_START` is different for `ramips`' subtarget:
* `0x1C000000` for `mt7620`
* `0x10000000` for `mt7621`

So this overflow is intentional.

See the [forum thread](https://forum.openwrt.org/t/hootoo-tm-05-add-lzma-loader/67687/308?u=xabolcs) for the details!
I planned to send a separate patch for that to ML when this PR got merged, as my pending D-Link DIR-860L patch also touches that value.
My plan was to drop these `FLASH_START` and `FLASH_END` defines and reintroduce the former as `CONFIG_FLASH_START` in `config.h`.

Reply:
> Hi @981213 , thank you for joining the review!
> 
> `FLASH_END` is unused in `loader.c`, but `FLASH_START` is different for `ramips`' subtarget:
> 
> * `0x1C000000` for `mt7620`
> * `0x10000000` for `mt7621`
It's not. 0x10000000 is the 256M offset of DRAM on mt7621.
(Google "mt7621 programming guide" and check  "Memory Map" section in the doc.)
It works for you probably because u-boot copied some data into ram.

> 
> So this overflow is intentional.

That's not the correct way to do it.
You should either make FLASH_START a build argument itself or define the correct value in loader.c according to some macros for SoC. (see UART_BASE in board-ralink.c for example.)

Reply:
> It's not. 0x10000000 is the 256M offset of DRAM on mt7621.
> (Google "mt7621 programming guide" and check "Memory Map" section in the doc.)
> It works for you probably because u-boot copied some data into ram.

Thanks for the details! I'll verify that for DIR-860L!

> That's not the correct way to do it.
> You should ...

I know. It was just a workaround to get the Loader work.

As this device would be the first device in `ramips` which uses the OKLI way of the Loader, I'm OK with any solution, as I had to verify and refresh my DIR-860L patch anyway.

Reply:
Yes, agreed - this was a bit of a dirty trick ... because folks were saying not to mess with FLASH_START. If it's allowed to change that, no issue here. It's easy 😄. Is there an agreement here? Again, no problem at all changing this, just want to make sure folks are on the same page.

And I agree with @xabolcs, thanks for the review comments!

Reply:
I think this is the only one left 😄. I can change the values, that's pretty easy, but some folks didn't want changes made to the loader code ... though it seems from the notes above I should make the change, this is somewhat a fixed value for the ramips devices, and not the same as AR71xx (which is what is used now)? Please just let me know.

Thanks!

Reply:
@arrmo 

> but some folks didn't want changes made to the loader code ...

@981213 is a [committer](https://openwrt.org/rules), and he provided his `Acked-by:` to the Loader change, so please choose your preferred solution and go for it:
> You should either make FLASH_START a build argument itself or define the correct value in `loader.c` according to some macros for SoC. (see `UART_BASE` in board-ralink.c for example.)

Reply:
About the values: see comment above!
> For the sake of correctness, you **should add a separated commit** fixing `FLASH_START` and `FLASH_END` to be `0x1c000000` and `0x1d000000` respectively, and define `FLASH_OFFS` as `0x200000` here.

Reply:
Will do! I'll align with what @981213 has noted - will change the values. And I'll change the names as well, it doesn't make sense to call out AR71xx here (yell if you disagree!).

Will update, send a bit later today. Stuck at work right now 🤣.

Reply:
OK, now that I get what you're after ... LOL. Done, and pushed as a separate commit, as you requested. The logic is much more clear now, thanks @981213, appreciate it!

I could be wrong, but I think this should address this comment (just yell if not). I did build and upgrade on the device - it all worked, settings saved and restored properly (and lzma-loader did jump straight to the correct address ... yay!).

Thanks for all the help and patience!

PS: As noted in the thread above, I did change FLASH_START, but dropped FLASH_END, as it's not used anywhere. Yell if you want that added in any case, I can do it easily.

Reply:
>  Done, and pushed as a separate commit, as you requested.

... with [formal issues](https://openwrt.org/submitting-patches)! :wink:

Reply:
Thanks for catching that! Dang signoff always catches me ... 🤣. Fixed!

Reply:
I linked it, because of:
* commit subject: missing `ramips:` prefix at least, but I would prefer to include `lzma-loader:` too (but I'm not a committer)
  * e.g. `ramips: lzma-loader: fix FLASH_START address`
* commit description is completely missing
* and yes, that sign-off too


:stuck_out_tongue_winking_eye:

Reply:
LOL - got it! I can change the subject, no issue. But not sure what else to add for the description - it's basically the same as the message, no?

Reply:
If somebody reads your commit she may or may not find out the original intent. Help the reader to understand your motivation!

Think about that this talk (here, in the PR) is not included in the commit!

For example
* why did you change `AR71XX_FLASH_START`'s value from `0x1f000000` to `0x1c000000`? It worked nicely with the old value! Wouldn't be better to change to `0x0`? Why is `0x1c000000` the value that we need here?
* why did you rename `AR71XX_FLASH_START` to `FLASH_START`?
* why did you remove `AR71XX_FLASH_END`?
* why did you keep `KSEG1ADDR(_x)` in `board-*.c`? It's unused too!

Reply:
Got it! Will add more detail, thanks.